dataland-batch-manager/src/main/kotlin/org/dataland/datalandbatchmanager/service/CompanyUploader.kt

```diff
@@ -15,34 +18,53 @@ import java.net.SocketTimeoutException
 @Service
 class CompanyUploader(
     @Autowired private val companyDataControllerApi: CompanyDataControllerApi,
+    @Autowired private val objectMapper: ObjectMapper,
 ) {
     companion object {
         const val MAX_RETRIES = 3
         const val UNAUTHORIZED_CODE = 401
-        const val FORBIDDEN_CODE = 403
     }
 
     private val logger = LoggerFactory.getLogger(javaClass)
 
-    /**
-     * Uploads a single Company to the dataland backend performing at most MAX_RETRIES retries.
-     * This function absorbs errors and logs them.
-     */
-    fun uploadSingleCompany(
-        companyInformation: CompanyInformation,
-    ) {
+    @Suppress("ReturnCount")
+    private fun checkForDuplicateIdentifierAndGetConflictingCompanyId(exception: ClientException): String? {
+        if (exception.statusCode != HttpStatus.BAD_REQUEST.value()) {
+            return null
+        }
+
+        val exceptionResponse = exception.response!! as ClientError<*>
+        val exceptionBodyString = exceptionResponse.body.toString()
+
+        val errorResponseBody = objectMapper.readTree(exceptionBodyString)
+        val firstError = errorResponseBody.get("errors").get(0)
+        if (firstError.get("errorType").textValue() != "duplicate-company-identifier") {
+            return null
+        }
+
+        val conflictingIdentifiers = firstError.get("metaInformation")
+        if (!conflictingIdentifiers.isArray || conflictingIdentifiers.size() != 1) {
+            return null
+        }
+
+        val conflictingIdentifier = conflictingIdentifiers.get(0)
+        val conflictingIdentifierType = conflictingIdentifier.get("identifierType").textValue()
+        if (conflictingIdentifierType != "Lei") {
+            return null
+        }
+
+        return conflictingIdentifier.get("companyId").textValue()
+    }
+
+    private fun retryCatchingErrors(function: () -> Unit) {
```

Comment:
A more specific name for `function` would be great, e.g. `functionToRetry`. Actually, I'm not too sure about this right now.